contracts/USM.sol

```diff
@@ -148,21 +179,149 @@ contract USM is IUSM, ERC20Permit, Delegable {
+        // Here we use the idea that for any b and 0 <= p <= 1, we can crudely approximate b**p by 1 + (b-1)p = 1 + bp - p.
+        // Eg: 0.6**0.5 pulls 0.6 "about halfway" to 1 (0.8); 0.6**0.25 pulls 0.6 "about 3/4 of the way" to 1 (0.9).
+        // So b**p =~ b + (1-p)(1-b) = b + 1 - b - p + bp = 1 + bp - p.
+        // (Don't calculate it as 1 + (b-1)p because we're using uints, b-1 can be negative!)
+        return WAD + mintBurnAdjustmentStored.wadMul(decayFactor) - decayFactor;
+    }
+
+    /**
+     * @notice The current fund-defund adjustment, equal to the stored value decayed by time since fundDefundAdjustmentTimestamp.
+     */
+    function fundDefundAdjustment() public override view returns (uint) {
+        uint numHalvings = WAD * (block.timestamp - fundDefundAdjustmentTimestamp) / BUY_SELL_ADJUSTMENTS_HALF_LIFE;
+        uint decayFactor = numHalvings.wadHalfExp(10);
+        return WAD + fundDefundAdjustmentStored.wadMul(decayFactor) - decayFactor;
+    }
+
+    /**
+     * @notice How much USM a minter currently gets back for ethIn ETH, accounting for adjustments and sliding prices.
+     *
+     * @param ethIn The amount of ETH passed to mint().
+     * @return The amount of USM, and the factor by which the ethPool grew (eg, 1.2 * WAD = 1.2x).
+     */
+    function usmFromMint(uint ethIn) public override view returns (uint, uint) {
+        // Mint USM at a sliding-up USM price (ie, at a sliding-down ETH price).  **BASIC RULE:** anytime pool_eth changes by
+        // factor k, eth_price changes by factor 1/k**2.  (Earlier versions of this logic scaled price by 1/k, not 1/k**2; but
+        // that results in calls to log()/exp().)
+        uint initialUsmPrice = usmPrice(Side.Buy);
+        uint _ethPool = ethPool();
+        uint ethPoolGrowthFactor = (_ethPool + ethIn).wadDiv(_ethPool);
+        // Math: this is an integral - sum of all USM minted at a sliding-down ETH price:
+        uint usmOut = _ethPool.wadDiv(initialUsmPrice).wadMul(WAD - WAD.wadDiv(ethPoolGrowthFactor));
+        return (usmOut, ethPoolGrowthFactor);
+    }
+
+    /**
+     * @notice How much ETH a burner currently gets from burning usmIn USM, accounting for adjustments and sliding prices.
+     *
+     * @param usmIn The amount of USM passed to burn().
+     * @return The amount of ETH, and the factor by which the ethPool shrank.
+     */
+    function ethFromBurn(uint usmIn) public override view returns (uint, uint) {
+        // Burn at a sliding-up price:
+        uint initialUsmPrice = usmPrice(Side.Sell);
+        uint _ethPool = ethPool();
+        // Math: this is an integral - sum of all USM burned at a sliding price.  Follows the same mathematical invariant as
+        // above: if pool_eth *= k, eth_price *= 1/k**2.
+        uint ethOut = WAD.wadDiv(WAD.wadDiv(_ethPool) + WAD.wadDiv(usmIn.wadMul(initialUsmPrice)));
+        uint ethPoolShrinkFactor = (_ethPool - ethOut).wadDiv(_ethPool);
```

Comment:
Shouldn't we check that `ethPool` >= `ethOut`?

Reply:
Ditto.  There should now be no remaining naked `+` or `-` ops.